File: python/523.py
import sys

sys.stdin = open("523.txt")

cases = int(sys.stdin.readline())
sys.stdin.readline()

for i in range(cases):
	if i > 0:
		print()
	matrix = []
	a = sys.stdin.readline().strip().split()
	n = len(a)
	matrix.append([])
	for j in a:
		if j == "-1":
			matrix[0].append(999999999)
		else:
			matrix[0].append(int(j))
	for j in range(1,n):
		matrix.append([])
		a = sys.stdin.readline().strip().split()
		for k in a:
			if k == "-1":
				matrix[j].append(999999999)
			else:
				matrix[j].append(int(k))
	taxes = []
	a = sys.stdin.readline().strip().split()
	for j in a:
		taxes.append(int(j))
	nexts = [[j for j in range(n)] for k in range(n)]
	for u in range(n):
		for s in range(n):
			if matrix[s][u] != 999999999:
				for t in range(n):
					if matrix[u][t] != 999999999 and matrix[s][t] > matrix[s][u] + taxes[u] + matrix[u][t]:
						matrix[s][t] = matrix[s][u] + taxes[u] + matrix[u][t]
						nexts[s][t] = nexts[s][u]
	printed = False
	while True:
		a = sys.stdin.readline().strip().split()
		if a == []:
			break

		if printed:
			print()
		else:
			printed = True

		print("From {0} to {1} :".format(a[0], a[1]))
		s = int(a[0]) - 1
		e = int(a[1]) - 1
		print("Path: {0}".format(s+1), end="")
		if s != e:
			u = s
			while u != e:
				u = nexts[u][e]
				print("-->{0}".format(u+1), end="")
		print()
		print("Total cost : {0}".format(matrix[s][e]))

File: c++/523.cc
#include <bits/stdc++.h>

using namespace std;

int main() 
{
	ifstream cin("523.txt");

	int cases, tmp;
	cin >> cases;

	string st;
	getline(cin, st);
	getline(cin, st);

	for (int i{0}; i<cases; ++i) {
		if (i > 0) {cout << endl;}
		vector<vector<int>> matrix;
		matrix.push_back(vector<int>{});
		getline(cin, st);
		stringstream ss(st);
		int n = 0;
		while(ss>>tmp) {
			if (tmp == -1) {matrix[0].push_back(999999999);}
			else {matrix[0].push_back(tmp);}
			++n;
		}
		for (int j{1}; j<n; ++j) {
			matrix.push_back(vector<int>{});
			getline(cin, st);
			stringstream ss(st);
			while(ss>>tmp) {
				if (tmp == -1) {matrix[j].push_back(999999999);}
				else {matrix[j].push_back(tmp);}	
			}
		}
		vector<int> taxes;
		getline(cin, st);
		stringstream ss2(st);
		while(ss2>>tmp) {
			taxes.push_back(tmp);
		}
		vector<vector<int>> nexts;
		for (int j{0}; j<n; ++j) {
			nexts.push_back(vector<int>{});
			for (int k{0}; k<n; ++k) {
				nexts[j].push_back(k);
			}
		}
		
		for (int u{0}; u<n; ++u) {
			for (int s{0}; s<n; ++s) {
				if (matrix[s][u] != 999999999) {
					for (int t{0}; t<n; ++t) {
						// cout << matrix[u][t] << " " << matrix[s][t] << " " << taxes[u] << " " << matrix[u][t] << endl;
						if (matrix[u][t] != 999999999 && matrix[s][t] > matrix[s][u] + taxes[u] + matrix[u][t]) {			
							matrix[s][t] = matrix[s][u] + taxes[u] + matrix[u][t];
							nexts[s][t] = nexts[s][u];
						}
					}
				}

			}
		}

		bool printed = false;
		while (true) {
			getline(cin, st);

			if (st == "") {break;}
			stringstream ss3(st);
			int s,e,u;
			ss3 >> s >> e;
			if (printed) {
				cout << endl;
			} else {
				printed = true;
			}
			--s;
			--e;
			cout << "From " << s+1 << " to " << e+1 << " :" << endl;
			cout << "Path: " << s+1;
			if (s != e) {
				u = s;
				while (u != e) {
					u = nexts[u][e];
					cout << "-->" << u+1;
				}
			}
			cout << endl;
			cout << "Total cost : " << matrix[s][e] << endl;
		}


	}

}